Deleted projects still returned by search api

Ive got Dependency Track v 4.4.2 set up with a MySQL db, and have been round a loop of importing projects and then deleting them - both vis the api. This has been in attempting to work out a parent/child relationship. Now, when I search with the API for a project, I get all the deleted instances - as shown by their UUIDs. For example, create projectX, version 1.0, 1.2 and 1.3. via the api. Next delete them via the api. Now, add them again via the api. Now, search for projectX via the api. I get six records, two for each version, each with different UUIDs. If I call GetProject with these UUIDs, I get a "project not found" error, and if I use the console, they're not shown.

Some 'interesting' things in the dptrack_apiserver.log file:

`2022-05-09 15:10:48,889 INFO [MetricsUpdateTask] Executing metrics update for project: 2b687196-4fa2-4b82-900a-3771b7c02864
2022-05-09 15:10:54,779 ERROR [VulnerabilityAnalysisTask] An unexpected error occurred performing a vulnerability analysis task
javax.jdo.JDOObjectNotFoundException: No such database row
	at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:563)
	at org.datanucleus.api.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1747)
	at alpine.persistence.AbstractAlpineQueryManager.detach(AbstractAlpineQueryManager.java:488)
	at org.dependencytrack.util.NotificationUtil.analyzeNotificationCriteria(NotificationUtil.java:63)
	at org.dependencytrack.tasks.scanners.BaseComponentAnalyzerTask.applyAnalysisFromCache(BaseComponentAnalyzerTask.java:97)
	at org.dependencytrack.tasks.scanners.OssIndexAnalysisTask.analyze(OssIndexAnalysisTask.java:161)
	at org.dependencytrack.tasks.scanners.OssIndexAnalysisTask.inform(OssIndexAnalysisTask.java:109)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.performAnalysis(VulnerabilityAnalysisTask.java:151)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.analyzeComponents(VulnerabilityAnalysisTask.java:117)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.inform(VulnerabilityAnalysisTask.java:69)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:99)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.datanucleus.exceptions.NucleusObjectNotFoundException: No such database row
	at org.datanucleus.store.rdbms.request.FetchRequest.execute(FetchRequest.java:439)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.fetchObject(RDBMSPersistenceHandler.java:316)
	at org.datanucleus.state.StateManagerImpl.loadFieldsFromDatastore(StateManagerImpl.java:1542)
	at org.datanucleus.state.StateManagerImpl.validate(StateManagerImpl.java:5521)
	at org.datanucleus.ExecutionContextImpl.findObject(ExecutionContextImpl.java:3562)
	at org.datanucleus.ExecutionContextImpl.findObject(ExecutionContextImpl.java:3021)
	at org.datanucleus.api.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1742)
	... 12 common frames omitted`
First occurred on 9th May, occurs many times mainly whilst deleting all the projects via the API.

Then, from 3:21pm on 9th May I see many, many of:
`2022-05-09 15:21:03,310 WARN [Persist] Delete of object "org.dependencytrack.model.Component@7307bec3" using statement "DELETE FROM "COMPONENT" WHERE "ID"=?" failed : Cannot delete or update a parent row: a foreign key constraint fails ("DependencyTrack"."DEPENDENCYMETRICS", CONSTRAINT "DEPENDENCYMETRICS_FK1" FOREIGN KEY ("COMPONENT_ID") REFERENCES "COMPONENT" ("ID"))
`
Then, this morning, there's a bunch of 
`2022-05-11 07:48:44,066 ERROR [Cache] >> EC.preCommit L1Cache op IS NULL!`


Does this point to anything?

--- a/src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
+++ b/src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
@@ -579,16 +579,18 @@ final class ProjectQueryManager extends QueryManager implements IQueryManager {
     /**
      * Deletes a Project and all objects dependant on the project.
      * @param project the Project to delete
+     * @param commitIndex specifies if the search index should be committed (an expensive operation)
      */
-    public void recursivelyDelete(Project project) {
+    public void recursivelyDelete(final Project project, final boolean commitIndex) {
         if (project.getChildren() != null) {
             for (final Project child: project.getChildren()) {
-                recursivelyDelete(child);
+                recursivelyDelete(child, false);
             }
         }
         pm.getFetchPlan().setDetachmentOptions(FetchPlan.DETACH_LOAD_FIELDS);
         final Project result = pm.getObjectById(Project.class, project.getId());
         Event.dispatch(new IndexEvent(IndexEvent.Action.DELETE, pm.detachCopy(result)));
+        commitSearchIndex(commitIndex, Project.class);
 
         deleteAnalysisTrail(project);
         deleteViolationAnalysisTrail(project);

--- a/src/main/java/org/dependencytrack/persistence/QueryManager.java
+++ b/src/main/java/org/dependencytrack/persistence/QueryManager.java
@@ -382,8 +382,8 @@ public class QueryManager extends AlpineQueryManager {
         return getProjectQueryManager().updateLastBomImport(p, date, bomFormat);
     }
 
-    public void recursivelyDelete(Project project) {
-        getProjectQueryManager().recursivelyDelete(project);
+    public void recursivelyDelete(final Project project, final boolean commitIndex) {
+        getProjectQueryManager().recursivelyDelete(project, commitIndex);
     }
 
     public ProjectProperty createProjectProperty(final Project project, final String groupName, final String propertyName,

--- a/src/main/java/org/dependencytrack/resources/v1/ProjectResource.java
+++ b/src/main/java/org/dependencytrack/resources/v1/ProjectResource.java
@@ -422,8 +422,8 @@ public class ProjectResource extends AlpineResource {
             final Project project = qm.getObjectByUuid(Project.class, uuid, Project.FetchGroup.ALL.name());
             if (project != null) {
                 if (qm.hasAccess(super.getPrincipal(), project)) {
-                    LOGGER.info("Project " + project.toString() + " deletion request by " + super.getPrincipal().getName());
-                    qm.recursivelyDelete(project);
+                    LOGGER.info("Project " + project + " deletion request by " + super.getPrincipal().getName());
+                    qm.recursivelyDelete(project, true);
                     return Response.status(Response.Status.NO_CONTENT).build();
                 } else {
                     return Response.status(Response.Status.FORBIDDEN).entity("Access to the specified project is forbidden").build();

--- a/src/main/java/org/dependencytrack/search/IndexManager.java
+++ b/src/main/java/org/dependencytrack/search/IndexManager.java
@@ -33,7 +33,6 @@ import org.apache.lucene.document.StringField;
 import org.apache.lucene.document.TextField;
 import org.apache.lucene.index.CorruptIndexException;
 import org.apache.lucene.index.DirectoryReader;
-import org.apache.lucene.index.IndexReader;
 import org.apache.lucene.index.IndexWriter;
 import org.apache.lucene.index.IndexWriterConfig;
 import org.apache.lucene.index.Term;
@@ -66,6 +65,7 @@ public abstract class IndexManager implements AutoCloseable {
     private static final Logger LOGGER = Logger.getLogger(IndexManager.class);
     private IndexWriter iwriter;
     private MultiFieldQueryParser qparser;
+    private DirectoryReader searchReader;
     private final IndexType indexType;
 
     /**
@@ -159,20 +159,22 @@ public abstract class IndexManager implements AutoCloseable {
     }
 
     /**
-     * Returns an {@link IndexSearcher}.
-     * <p>
-     * A new instance is created on every invocation. This is necessary because
-     * searchers operate on the state of the index they've been given upon construction.
-     * If the index changed, a new searcher must be created in order to have those changes
-     * being reflected in search results. See {@link IndexSearcher} documentation.
+     * Returns an {@link IndexSearcher} by opening the index directory first, if necessary.
      *
      * @return an {@link IndexSearcher}
      * @throws IOException when the index directory cannot be opened
      * @since 3.0.0
      */
-    protected IndexSearcher getIndexSearcher() throws IOException {
-        final IndexReader reader = DirectoryReader.open(getDirectory());
-        return new IndexSearcher(reader);
+    protected synchronized IndexSearcher getIndexSearcher() throws IOException {
+        if (searchReader == null) {
+            searchReader = DirectoryReader.open(getDirectory());
+        } else {
+            final var changedReader = DirectoryReader.openIfChanged(searchReader);
+            if (changedReader != null) {
+                searchReader = changedReader;
+            }
+        }
+        return new IndexSearcher(searchReader);
     }
 
     /**

--- a/src/main/java/org/dependencytrack/search/IndexManager.java
+++ b/src/main/java/org/dependencytrack/search/IndexManager.java
@@ -65,7 +65,6 @@ public abstract class IndexManager implements AutoCloseable {
 
     private static final Logger LOGGER = Logger.getLogger(IndexManager.class);
     private IndexWriter iwriter;
-    private IndexSearcher isearcher;
     private MultiFieldQueryParser qparser;
     private final IndexType indexType;
 
@@ -160,17 +159,20 @@ public abstract class IndexManager implements AutoCloseable {
     }
 
     /**
-     * Returns an IndexSearcher by opening the index directory first, if necessary.
-     * @return an IndexSearcher
+     * Returns an {@link IndexSearcher}.
+     * <p>
+     * A new instance is created on every invocation. This is necessary because
+     * searchers operate on the state of the index they've been given upon construction.
+     * If the index changed, a new searcher must be created in order to have those changes
+     * being reflected in search results. See {@link IndexSearcher} documentation.
+     *
+     * @return an {@link IndexSearcher}
      * @throws IOException when the index directory cannot be opened
      * @since 3.0.0
      */
     protected IndexSearcher getIndexSearcher() throws IOException {
-        if (isearcher == null) {
-            final IndexReader reader = DirectoryReader.open(getDirectory());
-            isearcher = new IndexSearcher(reader);
-        }
-        return isearcher;
+        final IndexReader reader = DirectoryReader.open(getDirectory());
+        return new IndexSearcher(reader);
     }
 
     /**

